src/Features/Core/Portable/Completion/Providers/TypeImportCompletionService.cs

```diff
@@ -0,0 +1,298 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal interface ITypeImportCompletionService : IWorkspaceService
+    {
+        /// <summary>
+        /// Get all the top level types from given project. This method is intended to be used for 
+        /// getting types from source only, so the project must support compilation. 
+        /// For getting types from PE, use <see cref="GetTopLevelTypesFromPEReference"/>.
+        /// </summary>
+        Task GetTopLevelTypesAsync(
+            Project project,
+            Action<TypeImportCompletionItemInfo> handleItem,
+            CancellationToken cancellationToken);
+
+        void GetTopLevelTypesFromPEReference(
+            Solution solution,
+            Compilation compilation,
+            PortableExecutableReference peReference,
+            Action<TypeImportCompletionItemInfo> handleItem,
+            CancellationToken cancellationToken);
+    }
+
+    internal readonly struct TypeImportCompletionItemInfo
+    {
+        public TypeImportCompletionItemInfo(CompletionItem item, bool isPublic)
+        {
+            Item = item;
+            IsPublic = isPublic;
+        }
+
+        public CompletionItem Item { get; }
+
+        public bool IsPublic { get; }
+    }
+
+    [ExportWorkspaceServiceFactory(typeof(ITypeImportCompletionService), ServiceLayer.Editor), Shared]
+    internal sealed class TypeImportCompletionService : IWorkspaceServiceFactory
+    {
+        private readonly ConcurrentDictionary<string, ReferenceCacheEntry> _peItemsCache
+            = new ConcurrentDictionary<string, ReferenceCacheEntry>();
+
+        private readonly ConcurrentDictionary<ProjectId, ReferenceCacheEntry> _projectItemsCache
```

Comment:
(FYI this caching mechanism here is *very* worrisome to me in general. I'm not hitting block but would want a lot of attention given here before we merge to make sure this isn't causing any sort of leaks.)

Reply:
yea, that's a good point. I should create a solution crawler so things got cleaned up as soon as possible (when type symbol might be changed because of source/config/etc. change).

Reply:
After discussion with @CyrusNajmabadi , I will started with listening to workspace events first.

Reply:
done
https://github.com/dotnet/roslyn/pull/34808/commits/d487b9664a40c49e912e127231d6427d0495c550